sklearn/externals/_lobpcg.py

```diff
@@ -356,28 +426,87 @@ def lobpcg(
         else:
             eigvals = (0, sizeX - 1)
 
-        A_dense = A(np.eye(n, dtype=A.dtype))
-        B_dense = None if B is None else B(np.eye(n, dtype=B.dtype))
+        try:
+            if isinstance(A, LinearOperator):
+                A = A(np.eye(n, dtype=int))
+            elif callable(A):
+                A = A(np.eye(n, dtype=int))
+                if A.shape != (n, n):
+                    raise ValueError(
+                        f"The shape {A.shape} of the primary matrix\n"
+                        f"defined by a callable object is wrong.\n"
+                    )
+            elif isspmatrix(A):
+                A = A.toarray()
+            else:
+                A = np.asarray(A)
+        except Exception as e:
+            raise Exception(
+                f"Primary MatMul call failed with error\n"
+                f"{e}\n")
 
-        vals, vecs = eigh(A_dense,
-                          B_dense,
-                          eigvals=eigvals,
-                          check_finite=False)
-        if largest:
-            # Reverse order to be compatible with eigs() in 'LM' mode.
-            vals = vals[::-1]
-            vecs = vecs[:, ::-1]
+        if B is not None:
+            try:
+                if isinstance(B, LinearOperator):
+                    B = B(np.eye(n, dtype=int))
+                elif callable(B):
+                    B = B(np.eye(n, dtype=int))
+                    if B.shape != (n, n):
+                        raise ValueError(
+                            f"The shape {B.shape} of the secondary matrix\n"
+                            f"defined by a callable object is wrong.\n"
+                        )
+                elif isspmatrix(B):
+                    B = B.toarray()
+                else:
+                    B = np.asarray(B)
+            except Exception as e:
+                raise Exception(
+                    f"Secondary MatMul call failed with error\n"
+                    f"{e}\n")
 
-        return vals, vecs
+        try:
+            if "subset_by_index" in inspect.signature(eigh).parameters:
+                # scipy >= 1.5
+                additional_params = {"subset_by_index": eigvals}
+            else:
+                # deprecated in scipy == 1.10
+                additional_params = {"eigvals": eigvals}
+            vals, vecs = eigh(A,
+                              B,
+                              check_finite=False,
+                              **additional_params)
+            if largest:
+                # Reverse order to be compatible with eigs() in 'LM' mode.
+                vals = vals[::-1]
+                vecs = vecs[:, ::-1]
+
+            return vals, vecs
+        except Exception as e:
+            raise Exception(
+                f"Dense eigensolver failed with error\n"
+                f"{e}\n"
+            )
```

Comment:
Is this try/except block is already in scipy as such, then so be it. But if not, then I am not sure why we need to try/except at all here. Re-raising an instance of the generic `Exception` class is weird. But if we want to do so, we can always forward the full original exception object as follows (which is more informative than just including the string message because the full traceback is preserved).

```suggestion
            raise Exception("LOBPCG dense eigensolver failed") from e
```

Reply:
This try/except is in SciPy. I just make the parameter dispatching for `eigvals` and `subset_by_index`

Reply:
ok